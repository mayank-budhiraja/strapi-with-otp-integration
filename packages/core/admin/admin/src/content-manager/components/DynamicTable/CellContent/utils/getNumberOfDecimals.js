export const getNumberOfDecimals = value => {
  if (value % 1 !== 0) {
    // value has decimals
    return value.toString().split('.')[1].length;
  }

  return 0;
};
