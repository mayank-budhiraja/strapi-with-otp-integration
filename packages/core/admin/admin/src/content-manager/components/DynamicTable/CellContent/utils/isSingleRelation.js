export default function isSingleRelation(type) {
  return ['oneToOne', 'manyToOne', 'oneToOneMorph'].includes(type);
}
