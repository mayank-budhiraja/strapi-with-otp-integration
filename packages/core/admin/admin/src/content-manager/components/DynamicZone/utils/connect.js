import React from 'react';

function connect(WrappedComponent, select) {
  return function(props) {
    // eslint-disable-next-line react/prop-types
    const selectors = select(props.name);

    return <WrappedComponent {...props} {...selectors} />;
  };
}

export default connect;
