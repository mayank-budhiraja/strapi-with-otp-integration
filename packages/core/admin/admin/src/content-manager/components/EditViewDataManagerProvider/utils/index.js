export { default as moveFields } from './moveFields';
export { default as cleanData } from './cleanData';
export { default as createYupSchema } from './schema';
