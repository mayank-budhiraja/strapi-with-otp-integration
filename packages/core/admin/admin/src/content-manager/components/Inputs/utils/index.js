export { default as connect } from './connect';
export { default as generateOptions } from './generateOptions';
export { default as getInputType } from './getInputType';
export { default as getStep } from './getStep';
export { default as select } from './select';
export { default as VALIDATIONS_TO_OMIT } from './VALIDATIONS_TO_OMIT';
