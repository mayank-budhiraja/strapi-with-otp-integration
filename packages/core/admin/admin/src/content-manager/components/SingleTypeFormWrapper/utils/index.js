// eslint-disable-next-line import/prefer-default-export
export { default as getRequestUrl } from './getRequestUrl';
