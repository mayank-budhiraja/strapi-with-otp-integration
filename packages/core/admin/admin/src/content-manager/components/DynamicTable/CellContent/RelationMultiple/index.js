import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useQuery } from 'react-query';
import { useIntl } from 'react-intl';
import { Typography } from '@strapi/design-system/Typography';
import { Box } from '@strapi/design-system/Box';
import { Badge } from '@strapi/design-system/Badge';
import { SimpleMenu, MenuItem } from '@strapi/design-system/SimpleMenu';
import { Loader } from '@strapi/design-system/Loader';
import styled from 'styled-components';
import { useNotifyAT } from '@strapi/design-system/LiveRegions';
import { stopPropagation } from '@strapi/helper-plugin';
import CellValue from '../CellValue';
import { axiosInstance } from '../../../../../core/utils';
import { getRequestUrl, getTrad } from '../../../../utils';

const TypographyMaxWidth = styled(Typography)`
  max-width: 500px;
`;

const fetchRelation = async (endPoint, notifyStatus) => {
  const {
    data: { results, pagination },
  } = await axiosInstance.get(endPoint);

  notifyStatus();

  return { results, pagination };
};

const RelationMultiple = ({ fieldSchema, metadatas, queryInfos, name, rowId, value }) => {
  const { formatMessage } = useIntl();
  const { notifyStatus } = useNotifyAT();
  const requestURL = getRequestUrl(`${queryInfos.endPoint}/${rowId}/${name.split('.')[0]}`);
  const [isOpen, setIsOpen] = useState(false);

  const Label = (
    <>
      <Badge>{value.count}</Badge>{' '}
      {formatMessage(
        {
          id: 'content-manager.containers.ListPage.items',
          defaultMessage: '{number, plural, =0 {items} one {item} other {items}}',
        },
        { number: value.count }
      )}
    </>
  );

  const notify = () => {
    const message = formatMessage({
      id: getTrad('DynamicTable.relation-loaded'),
      defaultMessage: 'Relations have been loaded',
    });
    notifyStatus(message);
  };

  const { data, status } = useQuery(
    [fieldSchema.targetModel, rowId],
    () => fetchRelation(requestURL, notify),
    {
      enabled: isOpen,
      staleTime: 0,
    }
  );

  return (
    <Box {...stopPropagation}>
      <SimpleMenu
        label={Label}
        size="S"
        onOpen={() => setIsOpen(true)}
        onClose={() => setIsOpen(false)}
      >
        {status !== 'success' && (
          <MenuItem aria-disabled>
            <Loader small>
              {formatMessage({
                id: getTrad('DynamicTable.relation-loading'),
                defaultMessage: 'Relations are loading',
              })}
            </Loader>
          </MenuItem>
        )}

        {status === 'success' && (
          <>
            {data?.results.map(entry => (
              <MenuItem key={entry.id} aria-disabled>
                <TypographyMaxWidth ellipsis>
                  <CellValue
                    type={metadatas.mainField.schema.type}
                    value={entry[metadatas.mainField.name] || entry.id}
                  />
                </TypographyMaxWidth>
              </MenuItem>
            ))}

            {data?.pagination.total > 10 && (
              <MenuItem
                aria-disabled
                aria-label={formatMessage({
                  id: getTrad('DynamicTable.relation-more'),
                  defaultMessage: 'This relation contains more entities than displayed',
                })}
              >
                <Typography>...</Typography>
              </MenuItem>
            )}
          </>
        )}
      </SimpleMenu>
    </Box>
  );
};

RelationMultiple.propTypes = {
  fieldSchema: PropTypes.shape({
    relation: PropTypes.string,
    targetModel: PropTypes.string,
    type: PropTypes.string.isRequired,
  }).isRequired,
  metadatas: PropTypes.shape({
    mainField: PropTypes.shape({
      name: PropTypes.string.isRequired,
      schema: PropTypes.shape({ type: PropTypes.string.isRequired }).isRequired,
    }),
  }).isRequired,
  name: PropTypes.string.isRequired,
  rowId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  queryInfos: PropTypes.shape({ endPoint: PropTypes.string.isRequired }).isRequired,
  value: PropTypes.object.isRequired,
};

export default RelationMultiple;
