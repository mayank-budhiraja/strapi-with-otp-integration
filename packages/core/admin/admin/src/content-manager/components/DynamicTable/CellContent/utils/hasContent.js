import isEmpty from 'lodash/isEmpty';
import isNumber from 'lodash/isNumber';

import isSingleRelation from './isSingleRelation';
import isFieldTypeNumber from '../../../../utils/isFieldTypeNumber';

export default function hasContent(type, content, metadatas, fieldSchema) {
  if (type === 'component') {
    const {
      mainField: { name: mainFieldName, type: mainFieldType },
    } = metadatas;

    // Repeatable fields show the ID as fallback, in case the mainField
    // doesn't have any content
    if (fieldSchema?.repeatable) {
      return content.length > 0;
    }

    const value = content?.[mainFieldName];

    // relations, media ... show the id as fallback
    if (mainFieldName === 'id' && ![undefined, null].includes(value)) {
      return true;
    }

    /* The ID field reports itself as type `integer`, which makes it
       impossible to distinguish it from other number fields.

       Biginteger fields need to be treated as strings, as `isNumber`
       doesn't deal with them.
    */
    if (
      isFieldTypeNumber(mainFieldType) &&
      mainFieldType !== 'biginteger' &&
      mainFieldName !== 'id'
    ) {
      return isNumber(value);
    }

    return !isEmpty(value);
  }

  if (type === 'relation') {
    if (isSingleRelation(fieldSchema.relation)) {
      return !isEmpty(content);
    }

    return content.count > 0;
  }

  /* 
      Biginteger fields need to be treated as strings, as `isNumber`
      doesn't deal with them.
  */
  if (isFieldTypeNumber(type) && type !== 'biginteger') {
    return isNumber(content);
  }

  if (type === 'boolean') {
    return content !== null;
  }

  return !isEmpty(content);
}
