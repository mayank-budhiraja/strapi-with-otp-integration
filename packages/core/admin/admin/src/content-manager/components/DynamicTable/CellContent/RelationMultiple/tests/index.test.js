import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { ThemeProvider, lightTheme } from '@strapi/design-system';
import { IntlProvider } from 'react-intl';
import { QueryClientProvider, QueryClient } from 'react-query';

import { axiosInstance } from '../../../../../../core/utils';
import RelationMultiple from '../index';

jest.spyOn(axiosInstance, 'get').mockResolvedValue({
  data: {
    results: [
      {
        id: 1,
        name: 'Relation entity 1',
      },
    ],

    pagination: {
      total: 1,
    },
  },
});

const DEFAULT_PROPS_FIXTURE = {
  fieldSchema: {
    type: 'relation',
    relation: 'manyToMany',
    target: 'api::category.category',
  },
  queryInfos: {
    endPoint: 'collection-types/api::address.address',
  },
  metadatas: {
    mainField: {
      name: 'name',
      schema: {
        type: 'string',
      },
    },
  },
  value: {
    count: 1,
  },
  name: 'categories.name',
  rowId: 1,
};

const ComponentFixture = () => {
  const queryClient = new QueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider theme={lightTheme}>
        <IntlProvider locale="en" messages={{}} defaultLocale="en">
          <RelationMultiple {...DEFAULT_PROPS_FIXTURE} />
        </IntlProvider>
      </ThemeProvider>
    </QueryClientProvider>
  );
};

describe('DynamicTabe / Cellcontent / RelationMultiple', () => {
  it('renders and matches the snapshot', async () => {
    const { container } = render(<ComponentFixture />);
    expect(container).toMatchSnapshot();
    expect(axiosInstance.get).toHaveBeenCalledTimes(0);
  });

  it('fetches relation entities once the menu is opened', async () => {
    const { container } = render(<ComponentFixture />);
    const button = container.querySelector('[type=button]');

    fireEvent(button, new MouseEvent('mousedown', { bubbles: true }));

    expect(screen.getByText('Relations are loading')).toBeInTheDocument();
    expect(axiosInstance.get).toHaveBeenCalledTimes(1);
    await waitFor(() => expect(screen.getByText('Relation entity 1')).toBeInTheDocument());
  });
});
