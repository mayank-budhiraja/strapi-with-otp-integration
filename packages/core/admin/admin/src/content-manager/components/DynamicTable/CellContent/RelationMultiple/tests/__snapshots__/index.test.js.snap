// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicTabe / Cellcontent / RelationMultiple renders and matches the snapshot 1`] = `
.c11 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c5 {
  background: #f6f6f9;
  padding: 4px;
  border-radius: 4px;
  min-width: 20px;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  color: #666687;
  font-weight: 600;
  font-size: 0.6875rem;
  line-height: 1.45;
  text-transform: uppercase;
}

.c4 {
  font-weight: 600;
  color: #32324d;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c9 {
  padding-left: 8px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  padding: 8px;
  border-radius: 4px;
  background: #ffffff;
  border: 1px solid #dcdce4;
  position: relative;
  outline: none;
}

.c0 svg {
  height: 12px;
  width: 12px;
}

.c0 svg > g,
.c0 svg path {
  fill: #ffffff;
}

.c0[aria-disabled='true'] {
  pointer-events: none;
}

.c0:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c0:focus-visible {
  outline: none;
}

.c0:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 16px;
  background: #4945ff;
  border: 1px solid #4945ff;
  border: 1px solid transparent;
  background: transparent;
}

.c1 .c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 .c3 {
  color: #ffffff;
}

.c1[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
}

.c1[aria-disabled='true'] .c3 {
  color: #666687;
}

.c1[aria-disabled='true'] svg > g,
.c1[aria-disabled='true'] svg path {
  fill: #666687;
}

.c1[aria-disabled='true']:active {
  border: 1px solid #dcdce4;
  background: #eaeaef;
}

.c1[aria-disabled='true']:active .c3 {
  color: #666687;
}

.c1[aria-disabled='true']:active svg > g,
.c1[aria-disabled='true']:active svg path {
  fill: #666687;
}

.c1:hover {
  background-color: #f6f6f9;
}

.c1:active {
  border: 1px solid undefined;
  background: undefined;
}

.c1 .c3 {
  color: #32324d;
}

.c1 svg > g,
.c1 svg path {
  fill: #8e8ea9;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c10 svg {
  height: 4px;
  width: 6px;
}

.c2 {
  padding: 4px 12px;
}

<div>
  <div
    aria-hidden="true"
    class=""
    role="button"
  >
    <div>
      <button
        aria-controls="simplemenu-1"
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="true"
        class="c0 c1 c2"
        type="button"
      >
        <span
          class="c3 c4"
        >
          <div
            class="c5 c6"
          >
            <span
              class="c7"
            >
              1
            </span>
          </div>
           
          item
        </span>
        <div
          aria-hidden="true"
          class="c8 c9"
        >
          <span
            class="c10"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="1em"
              viewBox="0 0 14 8"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M14 .889a.86.86 0 01-.26.625L7.615 7.736A.834.834 0 017 8a.834.834 0 01-.615-.264L.26 1.514A.861.861 0 010 .889c0-.24.087-.45.26-.625A.834.834 0 01.875 0h12.25c.237 0 .442.088.615.264a.86.86 0 01.26.625z"
                fill="#32324D"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </button>
    </div>
  </div>
  <div
    class="c11"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
